### ISSUE TITLE
Instabug app | Issue in authentication flow

### ISSUE THREAD

**Issue by AbdallahHemdan** [DataDog/apps#160]
Hey everyone, we already have an app for Instabug that already published almost 6 months ago and it was functioning very well and the whole flow of authentication, and rendering the app widget was working fine. Recently we got a report from a customer that he is not able to complete the login flow.

We already have the [Authentication](https://github.com/DataDog/apps/blob/master/docs/en/programming-model.md#authentication) feature enabled and have our authentication mechanism (email/password) based login.
![image](https://github.com/DataDog/apps/assets/40190772/8dada6f5-abb8-43eb-bcc0-cda72921ec94)

The logic behind the authentication we currently have is as follows:
1. Redirecting the user to the login page
2. User enters email/password
3. Once the logic success, we set the user to auth token to local storage to be used later for authorization in all requests used in the widget

The problem that currently happens is that the widget couldn't find the set auth token in the local storage even though it seems that it's already set in the original page for login which follows the suggested flow [here](https://github.com/DataDog/apps/blob/master/docs/en/programming-model.md#authentication:~:text=return%20getCookieValue(%22auth_token%22)%3B%0A%7D%3B-,authStateCallback,-%3A%20()%20%3D%3E%20%7B%0A%20%20return%20localStorage.getItem)

| Local storage in login page after login | Local storage in widget page in DataDog dashboard |
|------------------------|---------------------------|
| ![image](https://github.com/DataDog/apps/assets/40190772/e7091f5e-1ce0-457a-be57-37ddbcb1f49c) |  ![image](https://github.com/DataDog/apps/assets/40190772/a8568f19-cae4-4c2f-9fe2-c87c6851d3cf) |

It seems that there has been no change in our app implementation since we released and after debugging we couldn't find anything that causes the cleaning of the local storage, so my question:
1. Did you release any changes that may affect the sharing of the auth token or data in local storage?
2. Did you notice any problem that may cause this behavior?
3. Do you think there is something wrong with the auth flow or the technique of sharing the auth_token from the login to the widget to use this token later as an authorization token for the BE requests used there?

https://github.com/DataDog/apps/assets/40190772/ccbf05d5-5e30-4587-af35-fefd268f24ad

**Comment by felixge** [comment]
I work on profiling, so I'm unfamiliar with this part of our product and surprised by your ping 😅. That being said, I've forwarded this internally and hope somebody will be able to help.

**Comment by AbdallahHemdan** [comment]
@felixge Sorry for the incorrect mention, I was just recently an active contributor. Yes please, also let me know who is responsible for the apps part here 🙏🏻

**Comment by dsimoes-pd** [comment]
Hi. We are also experiencing the same issue with the authentication flow. 
The problem is due to a recent rollout from Google Chrome of [Storage Partitioning ](https://developer.chrome.com/en/docs/privacy-sandbox/storage-partitioning/). In short, Chrome is isolating the localStorage of the controller iframe from the window opened to perform login (even if they are on the same domain/origin).

Also, Firefox has applied standard privacy policy that disables cross site tracking.

**Comment by enbashi** [comment]
Hey, @AbdallahHemdan and @dsimoes-pd. I am a member of the apps team who's responsible for this product. We're looking into the issue and will get back to you soon with an update or a workaround if possible. Thanks for the report!

**Comment by iddl** [comment]
@dsimoes-pd @AbdallahHemdan we indeed confirm the release of [Storage Partitioning ](https://developer.chrome.com/en/docs/privacy-sandbox/storage-partitioning/) in Chrome 115 (release date: July 12th 2023) is preventing localStorage data to be shared between the login flow and the rest of the app code, this is clearly causing all of the login issues you are seeing.

We have two immediate solutions, and we're experimenting with a long term one. Both immediate solutions require changes on the PagerDuty and Instabug side, and we acknowledge it's a painful step for you. However, the long-term solution we are considering may not require any changes on your part. We will provide more information about this long-term solution as soon as we have confirmed is viable.

**Immediate solution 1: Abandoning localStorage in favor of cookies**

This would entail:
- Setting the user's access token in a cookie via your API's response on your login page
- The cookie should set **Secure** and **Same-Site: None attributes**
- If using **message** resolution, calling resolveAuthFlow({isAuthenticated: true}) in your login page after successful authentication and the access token cookie is set
- If using the default **polling** resolution, sending a request in **authStateCallback** that verifies whether a user's access token (stored in the request cookie) is authenticated

**Immediate solution 2: Implement the full login flow in the widget itself and still use localStoage**
- This requires [removing the authState callback object from the init call](https://github.com/DataDog/apps/blob/7a5dbfcba159f34188466520a9bdb672336bdf1e/examples/stream/admin/src/client.ts#L5-L14)
- It implies the login flow will visually appear in the widget, with a temporary subpar UX

If the details we have listed are unclear, we are more than happy to set up time to discuss with you.
We will send you direct updates on the long term solution as soon as we have verified its feasibility.
